fix(ReviewerServer): return scheduling states with customData

kuroahna opened the pull request:
## Purpose / Description

Currently, the `customData` that gets passed into the custom scheduler when [runStateMutationHook](https://github.com/ankidroid/Anki-Android/blob/v2.17alpha8/AnkiDroid/src/main/java/com/ichi2/anki/Reviewer.kt#L1087) is called is always empty.

According to the [SchedulingState](https://github.com/ankitects/anki/blob/23.10.1/proto/anki/scheduler.proto#L115-L118) protobuf message in Anki 23.10.1, it says
```protobuf
message SchedulingState {
  // ...
  oneof kind {
    Normal normal = 1;
    Filtered filtered = 2;
  }
  // The backend does not populate this field in GetQueuedCards; the front-end
  // is expected to populate it based on the provided Card. If it's not set when
  // answering a card, the existing custom data will not be updated.
  optional string custom_data = 3;
}
```

Hence, we need to populate this in the frontend

## Fixes
* Fixes https://github.com/ankidroid/Anki-Android/issues/14865

## Approach

When [getSchedulingStatesWithContext](https://github.com/ankidroid/Anki-Android/blob/v2.17alpha8/AnkiDroid/src/main/java/com/ichi2/anki/ReviewerServer.kt#L95) gets called in `ReviewerServer.kt`, we populate the `SchedulingStatesWithContext` protobuf message with the missing custom data. This `getSchedulingStatesWithContext()` method gets called before our custom JS scheduler gets executed when we call [mutateNextCardStates](https://github.com/ankitects/anki/blob/23.10.1/ts/reviewer/answering.ts#L45-L52)

Looking at Anki Desktop 23.10.1, they also populate the `states.current.customData` with `top_card.card.custom_data` in [reviewer.py](https://github.com/ankitects/anki/blob/23.10.1/qt/aqt/reviewer.py#L93-L99)

## How Has This Been Tested?

`./gradlew jacocoUnitTestReport` and `./gradlew jacocoAndroidTestReport` has passed on my machine with the emulator `Pixel_3a_API_34_extension_level_7_x86_64`

## Learning (optional, can help others)
Got [help](https://github.com/ankidroid/Anki-Android/issues/14865#issuecomment-1837692322) from @dae and avoided an unnecessary database call when retrieving the card's `customData`

## Checklist
_Please, go through these checks before submitting the PR._

- [x] You have a descriptive commit message with a short title (first line, max 50 chars).
- [x] You have commented your code, particularly in hard-to-understand areas
- [x] You have performed a self-review of your own code
- [ ] UI changes: include screenshots of all affected screens (in particular showing any new or changed strings)
- [ ] UI Changes: You have tested your change using the [Google Accessibility Scanner](https://play.google.com/store/apps/details?id=com.google.android.apps.accessibility.auditor)

kuroahna commented on AnkiDroid/src/main/java/com/ichi2/libanki/Card.kt:
```diff
@@ -130,20 +130,7 @@ open class Card : Cloneable {
-        oDue = card.originalDue.toLong()
-        oDid = card.originalDeckId
-        flags = card.flags
+        loadFromBackendCard(card)
```
This change was necessary because `customData` was not populated here in this constructor. I noticed some other fields were also not populated such as `mod`, `usn`, etc, so I re-used the `loadFromBackendCard(...)` method here

david-allison replied:
@BrayanDSO - pinging as you were in the area recently

BrayanDSO replied:
I expected something like this when I first read about issues with `customData`. 

I haven't changed it myself earlier because it wasn't related to the changes I were doing with the new Previewer.

This seems similar to what is done at `Cards.py` in Anki, so LGTM

mikehardy commented:
I also re-ran the test as it failed again - I believe it is the specific test in this PR though? Can't really merge a PR that has a flaky test right from the start 🤔

kuroahna commented on AnkiDroid/src/main/java/com/ichi2/libanki/Card.kt:
```diff
@@ -130,20 +130,7 @@ open class Card : Cloneable {
         renderOutput = null
```
Yeah, I was getting compiler errors. For `id`, `renderOutput`, and `note`, I was getting`Property must be initialized or abstract`. It's strange because we're calling `loadFromBackendCard(card)` which already _is_ initializing it. 

I looked online and they mentioned to use `lateinit` but I wasn't too comfortable with adding that since I'm not too familiar with Kotlin.

Adding `lateinit` to `id` didn't work because it says `lateinit modifier is not allowed on properties of primitive types` and needed to be converted to 
```kotlin
    var id by Delegates.notNull<Long>()
```

and 
```kotlin
    lateinit var renderOutput: TemplateRenderOutput?
    lateinit private var note: Note?
```
doesn't work because it says `lateinit is not allowed on properties of nullable types`

Is there a way to handle this?

david-allison replied:
That's fine, nothing more necessary, thanks for the explanation

kuroahna commented on AnkiDroid/src/androidTest/java/com/ichi2/anki/ReviewerTest.kt:
```diff
@@ -0,0 +1,146 @@
+        // the messages to be passed in and out of the WebView when evaluating
+        // the custom JS scheduler code. The card on the review screen takes
+        // some time to load, especially on an emulator
+        onWebView().withTimeout(10, TimeUnit.SECONDS).check(webContent(containingTextInBody(text)))
```
Yup makes sense, and ya, it's possible to remove the `withTimeout(...)` call. I actually added `withTimeout(10, TimeUnit.SECONDS)` here so that the code is self-documenting. When we call `onWebView()`, it actually initializes with a default timeout of 10 seconds

```java
    private WebInteraction(Matcher<View> viewMatcher) {
      this(viewMatcher, null, null, null, true, new Timeout(10, TimeUnit.SECONDS, true));
    }
```

The downside is that if they ever change the default timeout, we won't get it for free.

Also, we _do_ need the timeout, because that's how I ensure the card is fully loaded before clicking the `Show answer` and `Good` button. If we click `Good` too fast, then the custom scheduler code won't have enough time to execute and pass back the protobuf messages to the backend. Previously, I was using `Thread.sleep(1000)` but that was really unreliable and flaky.

I did look into [espresso idling resources](https://developer.android.com/training/testing/espresso/idling-resource) as an alternative which seems to be more reliable where we hook into the `onPageFinished` callback of the webview and set `isIdle` to `true` when it's done, but I couldn't get this to work because I had no way to access the `webview` that's created in `AbstractFlashcardViewer`

kuroahna commented:
> I've rerun the test to confirm whether it's flaky or broken. Emulator tests are regularly unstable.
> 
> We don't have good monitoring here, as we rarely use them. There are often timing or execution speed problems


> I also re-ran the test as it failed again - I believe it is the specific test in this PR though? Can't really merge a PR that has a flaky test right from the start 🤔

I was searching about `RootViewWithoutFocusException` and I found this issue https://github.com/robolectric/robolectric/issues/7428
> I searched on the GitHub about solution for RootViewWithoutFocusException, and **many people think it might be caused by slow Emulator on GitHub Actions. The Espresso waits only 10 seconds for RootView's focus state, and it might not enough for Emulators on GitHub Actions.**

As a quick and dirty hack, we could try increasing the `withTimeout(...)` from the default 10 seconds to something higher, but it's not ideal.

A potential, better long term solution would might be to use espresso idling resource (see [comment](https://github.com/ankidroid/Anki-Android/pull/14868#discussion_r1414151159)) and hook into the `onPageFinished` callback, but I had issues with getting access to the `webview` in `AbstractFlashcardViewer`

mikehardy commented:
🤔  I would be open to exposing the webview programmatically somehow, we have several other items that are exposed to non-private access for tests and as long as they are annotated correctly that works really well I think. You may have already tried that though - you appear to be quite resourceful at experimenting with various ways to make things work (which is awesome, by the way). *If* exposing the webview somehow so you could use espresso stuff would work, that would work for me.

If we just needed to have a larger timeout for this one test to make it work, I would be okay with that as well. Our emulator tests usually complete just a little before the unit test run on windows if I recall correctly. And either way, I would rather have a stable but slightly slower test then either of no test or unstable test

david-allison commented:
Very open to making anything we need public + `VisibleForTesting`. 

If we're increasing the timeout for testing, then I support it, ideally with the goal to reduce the test time.

If we can't get it working within a reasonable time period, I'd prefer to mark it as `@Flaky(OS = All)`.

It passes on my local machine, I'm happy with the code and it moves the project forwards

mikehardy commented:
Yeah - for the avoidance of doubt and because I have not stated it: this is awesome, and I'm also +1 on "whatever works" to get it moving

kuroahna commented:
> If we're increasing the timeout for testing, then I support it, ideally with the goal to reduce the test time later

> Yeah - for the avoidance of doubt and because I have not stated it: this is awesome, and I'm also +1 on "whatever works" to get it moving

I've pushed with a change to increase the timeout from 10 seconds to 2 minutes (being a little generous here for the CI build). Hopefully this passes on the CI build. If not, this will need some further investigation

>  I would be open to exposing the webview programmatically somehow, we have several other items that are exposed to non-private access for tests and as long as they are annotated correctly that works really well I think. You may have already tried that though

Sounds good! I actually didn't dive too deep into this as I didn't want to modify our impls just to allow testing, as I find that's usually a bad practice, but for this case, it does make sense. I'll spend some time later and see if I can get that to work. I would like to first see if the 2 minute timeout is enough to get the CI build to pass at least. And if we're in a rush to get this out, then I can make a follow up PR and see if I can refactor it to use espresso idling resource instead

mikehardy commented:
Extra timeout wasn't a fix unfortunately:

```

com.ichi2.anki.ReviewerTest > testCustomSchedulerWithCustomData[test(AVD) - 11] FAILED 
	androidx.test.espresso.base.RootViewPicker$RootViewWithoutFocusException: Waited for the root of the view hierarchy to have window focus and not request layout for 10 seconds. If you specified a non default root matcher, it may be picking a root that never takes focus. Root:
	Root{application-window-token=android.view.ViewRootImpl$W@b61fc3f, window-token=android.view.ViewRootImpl$W@b61fc3f, has-window-focus=false, layout-params-type=1, layout-params-string={(0,0)(fillxfill) ty=BASE_APPLICATION wanim=0x10302fe
```

kuroahna commented:
Hmm, seems like that 10 second timeout isn't actually configurable for the [RootViewPicker](https://github.com/android/android-test/blob/main/espresso/core/java/androidx/test/espresso/base/RootViewPicker.java#L104)

```java
  /**
   * Waits for a root to be ready. Ready here means the UI is no longer in flux if layout of the
   * root view is not being requested and the root view has window focus or is focusable.
   */
  private Root waitForRootToBeReady(Root pickedRoot) {
    long timeout = System.currentTimeMillis() + TimeUnit.SECONDS.toMillis(10) /* 10 seconds */;
    BackOff rootReadyBackoff = new RootReadyBackoff();
    while (System.currentTimeMillis() <= timeout) {
      if (pickedRoot.isReady()) {
        return pickedRoot;
      } else {
        controlledLooper.simulateWindowFocus(pickedRoot.getDecorView());
        uiController.loopMainThreadForAtLeast(rootReadyBackoff.getNextBackoffInMillis());
      }
    }

    throw new RootViewWithoutFocusException(
        String.format(
            Locale.ROOT,
            "Waited for the root of the view hierarchy to have "
                + "window focus and not request layout for 10 seconds. If you specified a non "
                + "default root matcher, it may be picking a root that never takes focus. "
                + "Root:\n%s",
            pickedRoot));
  }
```

I've temporarily used `Thread.sleep()` for 2 minutes to see if waiting long enough is actually the solution for the CI builds to pass. Definitely don't want to sleep for 2 minutes here.

kuroahna commented:
Okay, so that `Thread.sleep()` for 2 minutes didn't work. I think it's getting stuck at some other place. In the stacktrace, it shows:

```
12-04 18:35:01.098  5699  5721 E TestRunner: failed: testCustomSchedulerWithCustomData(com.ichi2.anki.ReviewerTest)
12-04 18:35:01.098  5699  5721 E TestRunner: ----- begin exception -----
...
12-04 18:35:01.099  5699  5721 E TestRunner: 	at com.ichi2.anki.ReviewerTest.clickOnDeckWithName(ReviewerTest.kt:101)
12-04 18:35:01.099  5699  5721 E TestRunner: 	at com.ichi2.anki.ReviewerTest.testCustomSchedulerWithCustomData(ReviewerTest.kt:71)
```

Seems like it can't click on the deck with the `RecyclerView`

```kotlin
    private fun clickOnDeckWithName(deckName: String) {
        onView(withId(R.id.files)).perform(
            RecyclerViewActions.actionOnItem<RecyclerView.ViewHolder>(
                hasDescendant(withText(deckName)),
                click()
            )
        )
    }
```

kuroahna commented:
Debugging the logs, I see that the DeckPicker activity is created -> IntroductionActivity created -> DeckPicker activity destroyed immediately after, which is strange
```
12-04 22:57:31.269  3741  3782 I TestRunner: started: testCustomSchedulerWithCustomData(com.ichi2.anki.ReviewerTest)
...
12-04 22:57:32.595  3741  3741 D LifecycleMonitor: Lifecycle status change: com.ichi2.anki.DeckPicker@492aa9b in: PRE_ON_CREATE
...
12-04 22:57:32.818  3741  3741 D LifecycleMonitor: Lifecycle status change: com.ichi2.anki.DeckPicker@492aa9b in: CREATED
...
12-04 22:57:32.907  3741  3741 D LifecycleMonitor: Lifecycle status change: com.ichi2.anki.IntroductionActivity@841927b in: PRE_ON_CREATE
...
12-04 22:57:33.156  3741  3741 I AnkiDroidApp$FragmentLifecycleLogger: IntroductionActivity::SetupCollectionFragment::onAttach
...
12-04 22:57:33.177  3741  3741 I AnkiDroidApp$FragmentLifecycleLogger: IntroductionActivity::SetupCollectionFragment::onCreate
12-04 22:57:33.179  3741  3741 D LifecycleMonitor: Lifecycle status change: com.ichi2.anki.IntroductionActivity@841927b in: CREATED
12-04 22:57:33.183  3741  3741 I AnkiDroidApp$onCreate: IntroductionActivity::onStart
12-04 22:57:33.245  3741  3741 I AnkiDroidApp$FragmentLifecycleLogger: IntroductionActivity::SetupCollectionFragment::onViewCreated
12-04 22:57:33.246  3741  3741 I AnkiDroidApp$FragmentLifecycleLogger: IntroductionActivity::SetupCollectionFragment::onStart
12-04 22:57:33.272  3741  3741 D LifecycleMonitor: Lifecycle status change: com.ichi2.anki.IntroductionActivity@841927b in: STARTED
12-04 22:57:33.274  3741  3741 I AnkiDroidApp$onCreate: IntroductionActivity::onResume
12-04 22:57:33.274  3741  3741 D UsageAnalytics: sendAnalyticsScreenView(): IntroductionActivity
12-04 22:57:33.274  3741  3741 D UsageAnalytics: getOptIn() status: false
12-04 22:57:33.275  3741  3741 D LifecycleMonitor: Lifecycle status change: com.ichi2.anki.IntroductionActivity@841927b in: RESUMED
12-04 22:57:33.276  3741  3741 I AnkiDroidApp$FragmentLifecycleLogger: IntroductionActivity::SetupCollectionFragment::onResume
...
12-04 22:57:33.607  3741  3741 D LifecycleMonitor: Lifecycle status change: com.ichi2.anki.DeckPicker@492aa9b in: DESTROYED
```

I tried launching `IntroductionActivity` instead of `DeckPicker` with

```kotlin
        ActivityScenario.launch(IntroductionActivity::class.java).use { 
// ...            
        }
```

And the [build to pass](https://github.com/ankidroid/Anki-Android/actions/runs/7093981850).

I also tried

```kotlin
    @get:Rule
    val activityScenarioRule = ActivityScenarioRule(IntroductionActivity::class.java)
```

but the [build](https://github.com/ankidroid/Anki-Android/actions/runs/7095544134) strangely failed 🤔 

Then I ran the build again, and it [passed](https://github.com/ankidroid/Anki-Android/actions/runs/7095958803/job/19313698798) with no actual changes (just did a rebase and squashed all commits and added a comment for why we're launching `IntroductionActivity` instead of `DeckPicker`). So that's weird...

I've noticed `DeckPickerTest` also launches the `DeckPicker` activity and has the test ignored out:
```kotlin
    @Ignore("This test appears to be flaky everywhere")
    @Test
    fun checkIfClickOnCountsLayoutOpensStudyOptionsOnMobile() {
...
    }

    @Test
    fun checkIfStudyOptionsIsDisplayedOnTablet() {
        // Run the test only on emulator.
        assumeTrue(isEmulator())
        assumeFalse("Test flaky in CI - #9282, skipping", wasBuiltOnCI())
...
    }
```

I wonder if these flaky tests have to do with launching `DeckPicker` instead of `IntroductionActivity`?

Also, I tried using the Espresso idling resource, in particular [CountingIdlingResource](https://developer.android.com/reference/androidx/test/espresso/idling/CountingIdlingResource), but it was hard to get a reference to the `Reviewer` activity since we launch the `IntroductionActivity`. There's the [onActivity](https://developer.android.com/reference/androidx/test/core/app/ActivityScenario#onActivity(androidx.test.core.app.ActivityScenario.ActivityAction%3CA%3E)) method in `ActivityScenario`, but it only gives me a reference to the `IntroductionActivity` and not `Reviewer` which creates the webview. 

I've also tried setting the webview id in `AbstractFlashcardViewer::createWebView` with
```kotlin
        webView.id = 123
```
and in my test 

```
        ActivityScenario.launch(IntroductionActivity::class.java).use { scenario ->
            scenario.onActivity { activity ->
                val webView = activity.findViewById<WebView>(123)
// ...
            }
        }
```

but I get
```
java.lang.NullPointerException: Cannot run onActivity since Activity has been destroyed already
```

Not sure if there's a way to do this?

In either case, I think waiting for the card to fully load with `onWebView().check(webContent(containingTextInBody(text)))` should be sufficient for now

mikehardy commented on AnkiDroid/src/androidTest/java/com/ichi2/anki/ReviewerTest.kt:
```diff
@@ -0,0 +1,143 @@
+@RunWith(AndroidJUnit4::class)
+class ReviewerTest : InstrumentedTest() {
+
+    // Launch IntroductionActivity instead of DeckPicker activity because in CI
+    // builds, it seems to create IntroductionActivity after the DeckPicker,
+    // causing the DeckPicker activity to be destroyed. As a consequence, this
+    // will throw RootViewWithoutFocusException when Espresso tries to interact
+    // with an already destroyed activity. By launching IntroductionActivity, we
+    // ensure that IntroductionActivity is launched first and navigate to the
+    // DeckPicker -> Reviewer activities
```
What an odyssey to get to this point: Thank you for pursuing it and I think this comment captures the new wisdom very well

david-allison reviewed (approved):
Thank you SO MUCH for taking the time to get the tests working & documented

You've removed a huge, time-consuming barrier for all future contributors. This will have an impact for years to come

Truly, kudos to you. Thank you!

mikehardy commented:
> It'll be really helpful if we can get a screenshot when the failure happens, as that'll help confirm whether or not this is true. Is it possible to get that setup for the emulator tests?

The last time I looked into doing that it did work but generated it's own overhead which slowed things down - and then tests were running smoothly enough with some other CI performance tweaks that I did not pursue it again

The modification would be in the yaml somewhere around here https://github.com/ankidroid/Anki-Android/blob/b1cb24ca85a2ca92a5fd4189e2eb43963a68e804/.github/workflows/tests_emulator.yml#L178-L180

https://developer.android.com/studio/run/emulator-record-screen

> adb emu screenrecord start --time-limit 10 [path to save video]/sample_video.webm

You could probably make the screenrecord timeline just some huge number, and then just `killall -HUP <name of process, maybe adb?>` whatever the spawned process appears to be for the screenrecord feature after the test step finishes?

Then upload artifact similar to the way we upload the logcat (although I have since learned that upload artifact already .zips the thing so my manual gzip of the log is redundant and no need to do the same to the screen record)

kuroahna commented:
According to [google's issue tracker](https://issuetracker.google.com/issues/236721231), the screen record is limited to a max of 3 minutes. In API 34, it's been increased to 30 minutes. I think we're still on api level 30

kuroahna commented:
I've opened a PR to record the screen here: https://github.com/ankidroid/Anki-Android/pull/14880

Seems like it's just stuck with these dialogs:

![image](https://github.com/ankidroid/Anki-Android/assets/85209455/57c34c12-5383-4f29-ad58-1d1f6e9f6d25)

![image](https://github.com/ankidroid/Anki-Android/assets/85209455/3cbd6763-45af-4f6a-b298-895b64646469)

![image](https://github.com/ankidroid/Anki-Android/assets/85209455/096201d9-b0f9-4ee6-848f-79824780c357)

mikehardy commented:
This is a job for 🥁 🥁  the test butler! 🤵‍♂️ https://github.com/linkedin/test-butler
